src/hotspot/share/opto/escape.cpp

```diff
@@ -480,28 +486,107 @@ bool ConnectionGraph::can_reduce_phi_check_inputs(PhiNode* ophi) const {
+      SafePointScalarMergeNode* smerge = sfpt_in->as_SafePointScalarMerge();
+      Node* nsr_ptr = sfpt->in(smerge->merge_pointer_idx(jvms));
+      if (nsr_ptr == n) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+// Check if we are able to untangle the merge. The following patterns are
+// supported:
+//  - Phi -> SafePoints
+//  - Phi -> CmpP/N
+//  - Phi -> AddP -> Load
+//  - Phi -> CastPP -> SafePoints
+//  - Phi -> CastPP -> AddP -> Load
+bool ConnectionGraph::can_reduce_check_users(Node* n, uint nesting) const {
+  for (DUIterator_Fast imax, i = n->fast_outs(imax); i < imax; i++) {
+    Node* use = n->fast_out(i);
 
     if (use->is_SafePoint()) {
-      if (use->is_Call() && use->as_Call()->has_non_debug_use(ophi)) {
-        NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d on invocation %d. Call has non_debug_use().", ophi->_idx, _invocation);)
+      if (use->is_Call() && use->as_Call()->has_non_debug_use(n)) {
+        NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d on invocation %d. Call has non_debug_use().", n->_idx, _invocation);)
+        return false;
+      } else if (has_been_reduced(n->is_Phi() ? n->as_Phi() : n->as_CastPP()->in(1)->as_Phi(), use->as_SafePoint())) {
+        NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d on invocation %d. It has already been reduced.", n->_idx, _invocation);)
         return false;
       }
     } else if (use->is_AddP()) {
       Node* addp = use;
       for (DUIterator_Fast jmax, j = addp->fast_outs(jmax); j < jmax; j++) {
         Node* use_use = addp->fast_out(j);
+        const Type* load_type = _igvn->type(use_use);
+
         if (!use_use->is_Load() || !use_use->as_Load()->can_split_through_phi_base(_igvn)) {
-          NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d on invocation %d. AddP user isn't a [splittable] Load(): %s", ophi->_idx, _invocation, use_use->Name());)
+          NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d on invocation %d. AddP user isn't a [splittable] Load(): %s", n->_idx, _invocation, use_use->Name());)
           return false;
+        } else if (nesting > 0 && load_type->isa_narrowklass()) {
+          NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d on invocation %d. Nested NarrowKlass Load: %s", n->_idx, _invocation, use_use->Name());)
+          return false;
+        }
+      }
+    } else if (nesting > 0) {
+      NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d on invocation %d. Unsupported user %s at nesting level %d.", n->_idx, _invocation, use->Name(), nesting);)
+      return false;
+    } else if (use->is_CastPP()) {
+      const Type* cast_t = _igvn->type(use);
+      if (cast_t == nullptr || cast_t->make_ptr()->isa_instptr() == nullptr) {
+        NOT_PRODUCT(use->dump();)
```

Comment:
This dump should be guarded by TraceReduceAllocationMerges as should the one at line 574

Reply:
I opened https://bugs.openjdk.org/browse/JDK-8330850 for this.

Reply:
Sorry I didn't include this here but I filed https://bugs.openjdk.org/browse/JDK-8330277 for it.